### PR TITLE
[GPU] Custom layout for onednn weights reorder

### PR DESCRIPTION
**Pull request by jade-cho**
Co-Authored-By: Jade Cho <jade.cho@intel.com>

+ Add a format_tags.
+ Add a function to check shape consistent for weight reorder input and output.

### Tickets:
 - *134081*
 - *132916*
 - *130599*
 - *119476*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp [hunk 1]
```diff
@@ -13,6 +13,25 @@
 namespace cldnn {
 namespace onednn {
 
+namespace {
+/*
+std::string convert_data_format_string(cldnn::format fmt) {
```
Comment (hyunback) [inline src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:18]: It seems dummy comment, Is it intended?

**Review by hyunback — APPROVED** [review]

**Review by e-ddykim — APPROVED** [review]
Looks good to me

File: src/plugins/intel_gpu/include/intel_gpu/runtime/format.hpp [hunk 2]
```diff
@@ -380,46 +387,48 @@ struct format {
...
-    const std::vector<uint64_t>& dims_order() const { return traits(value)._order; }
+    const std::vector<uint64_t>& dims_order() const { return traits()._order; }
     /// @brief Returns an order of dimensions in form of string.
-    const std::string& order() const { return traits(value).order; }
+    const std::string& order() const { return traits().order; }
     /// @brief Returns an internal orders of dimensions form of string.
-    const std::string& internal_order() const { return traits(value).internal_order; }
+    const std::string& internal_order() const { return traits().internal_order; }
+    /// @brief Returns an order of dimensions for a @ format.
+    const std::vector<size_t>& axes_order() const { return traits()._order; }
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/include/intel_gpu/runtime/format.hpp:417]: This is a duplication of `const std::vector<uint64_t>& dims_order() const { return traits()._order; }`

**Review by sshlyapn — APPROVED** [review]
Looks good to me, left minor comment

File: src/plugins/intel_gpu/include/intel_gpu/runtime/format.hpp (hunk 2 above)
Comment (jade-cho) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/format.hpp]: Removed

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp (hunk 1 above)
Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp]: Removed

File: src/plugins/intel_gpu/src/graph/impls/onednn/primitive_onednn_base.h [hunk 3]
```diff
@@ -482,6 +482,31 @@ struct typed_primitive_onednn_impl : public typed_primitive_impl<PType> {
         return args;
     }
 
+    virtual std::unordered_map<int, dnnl::memory> get_arguments(typed_primitive_inst<PType>& instance, kernel_arguments_data& mem_args) const {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/impls/onednn/primitive_onednn_base.h:485]: I think this is only for reorder. What about adding an assertion that checks `inputs.size() == 1` and `outputs.size() == 1`?

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp [hunk 4]
```diff
@@ -106,13 +106,46 @@ dnnl::memory::data_type convert_data_type(cldnn::data_types dt) {
...
+        { cldnn::format::giozyx, dnnl::memory::format_tag::giodhw },
+        { cldnn::format::goizyx,  dnnl::memory::format_tag::goidhw },
+
+        { cldnn::format::os_iyx_osv16,  dnnl::memory::format_tag::Oihw16o },
+        { cldnn::format::gs_oiyx_gsv8,  dnnl::memory::format_tag::Goihw8g },
+        { cldnn::format::gs_oiyx_gsv16,  dnnl::memory::format_tag::Goihw16g },
+        { cldnn::format::gs_oiyx_gsv32,  dnnl::memory::format_tag::Goihw32g },
+        { cldnn::format::gs_oizyx_gsv16,  dnnl::memory::format_tag::Goidhw16g },
+        { cldnn::format::gs_oizyx_gsv32,  dnnl::memory::format_tag::Goidhw32g },
+        { cldnn::format::g_os_iyx_osv16,  dnnl::memory::format_tag::gOihw16o },
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:131]: why do we need such formats here?

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 5]
```diff
@@ -1567,10 +1567,14 @@ impl_types layout_optimizer::get_preferred_impl_type(program_node& node, format
...
 
+        if (output_fmt == format::custom) {
+            return impl_types::onednn;
+        }
+
         preferred_impl = impl_types::onednn;
 
-        if (std::find(onednn_optimized_fmt.begin(), onednn_optimized_fmt.end(), input_fmt) == onednn_optimized_fmt.end() ||
-            std::find(onednn_optimized_fmt.begin(), onednn_optimized_fmt.end(), output_fmt) == onednn_optimized_fmt.end()) {
+        if ((std::find(onednn_optimized_fmt.begin(), onednn_optimized_fmt.end(), input_fmt) == onednn_optimized_fmt.end() ||
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1576]: nit: please remove additional parenthesis

**Review by isanghao — COMMENTED** [review]
looks good to me in general. some minor comments are left.

File: src/plugins/intel_gpu/src/graph/impls/onednn/convolution_onednn.cpp [hunk 6]
```diff
@@ -142,14 +143,25 @@ struct convolution_onednn : typed_primitive_onednn_impl<convolution> {
...
-
-        auto output_weights_layout = input_weights_layout;
-        output_weights_layout.format = out_fmt;
-
-        return std::make_shared<WeightsReorderParams>(input_weights_layout, output_weights_layout, rotate, grouped_weights);
+        auto source_weights_layout = impl_params.get_input_layout(1);
+        auto grouped_weights = format::is_grouped(source_weights_layout.format) || cldnn_prim->grouped_weights_shape;
+        auto target_weights_desc = pd.weights_desc(0);
+        OPENVINO_ASSERT(onednn::keep_weights_reorder_shape_consistent(source_weights_layout, target_weights_desc),
+                        "[GPU] Input shape and output shape of weight reorder should be same.");
```
[the remark this answers is not in the document]
Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/impls/onednn/convolution_onednn.cpp:150]: If it doesn't pass this function, need to assertion. Does this mean that another variable needs to be added?

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp (hunk 4 above)
Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:131]: Actually, they don't need. these weights formats are maybe used in unit tests. So, I added a few more in advance.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 5 above)
Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1576]: Will remove

File: src/plugins/intel_gpu/src/graph/graph_optimizer/propagate_constants.cpp [hunk 7]
```diff
@@ -5,10 +5,16 @@
...
 #include "program_node.h"
 #include "intel_gpu/runtime/engine.hpp"
+#include "intel_gpu/runtime/debug_configuration.hpp"
 #include "intel_gpu/graph/program.hpp"
 #include "intel_gpu/graph/network.hpp"
 #include "data_inst.h"
 #include "intel_gpu/runtime/itt.hpp"
+#ifdef ENABLE_ONEDNN_FOR_GPU
+#include "reorder_inst.h"
+#include "permute_inst.h"
```
[the remark this answers is not in the document]
Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/propagate_constants.cpp:15]: will remove

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp (hunk 4 above)
Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:131]: There are only a few formats to remove, so I'll leave them as is.

File: src/plugins/intel_gpu/src/graph/impls/onednn/primitive_onednn_base.h (hunk 3 above)
Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/impls/onednn/primitive_onednn_base.h:485]: Add assert lines for that.

**Comment by isanghao** [comment]
Performance was good on daily test

**Review by vladimir-paramuzov — APPROVED** [review]

**Pull request merged**